### DISCUSSION TITLE
How to reuse a `$lib` outside of the SvelteKit build?

### DISCUSSION THREAD

**Discussion by maximedupre** [sveltejs/kit#10985] [Q&A] [answered]
Let's say that at the root of my project I have a "./scripts" folder.

In this "./scripts" folder, I have a bunch of one-off scripts that I run via `ts-node` (e.g. `npx ts-node --esm ./scripts/clear-db`).

For instance, "./scripts/clear-db", could import one of my `$lib` helpers: `import { init } from '$lib/server/db';`, but that would give the following error:

> CustomError: Cannot find package '$lib' imported from .../scripts/clear-db.ts

This could be easily fixed by simply specifying the full relative path to the module: `import { init } from './src/lib/server/db';`

But then the problem is that if "./src/lib/server/db" itself, or any other file that it imports, or that those other files import, use a `$lib` import (or a `$env` import), it will cause the same issue.

It seems like the only way to not have this issue is to never use `$lib` and `$env` imports in the SvelteKit project.

What's the best way to reuse a `$lib` outside of the SvelteKit build?

Thanks for the help 🙏

**Comment by david-plugge** [comment, marked as answer]
You can use vite-node

**Reply by maximedupre** [reply]
Thanks David, that's actually a great idea. 

However, it seems like files outside the usual projects are treated as client-side code 🤔

e.g. `npx vite-node scripts/clear-db.ts`

Output 👇
```
file:///.../node_modules/@sveltejs/kit/src/exports/vite/index.js:358
                                        throw new Error(`Cannot import ${relative} into client-side code`);
                                              ^

Error: Cannot import $env/static/private into client-side code
    at Context.load (file:///.../node_modules/@sveltejs/kit/src/exports/vite/index.js:358:12)
    at Object.load (file:///.../node_modules/vite/dist/node/chunks/dep-4d3eff22.js:42892:46)
    at loadAndTransform (file:///.../node_modules/vite/dist/node/chunks/dep-4d3eff22.js:53318:24)

Node.js v18.13.0
```

I wonder if there is a way to make my "scripts/*.ts" be considered server-side by the Svelte Kit compiler.

**Reply by maximedupre** [reply]
I looked at "/node_modules/@sveltejs/kit/src/exports/vite/index.js" and found a not-so-elegant hack 😅

```
if (options?.ssr === false && process.env.TEST !== 'true') {
	const normalized_cwd = vite.normalizePath(cwd);
	const normalized_lib = vite.normalizePath(kit.files.lib);
	if (
		is_illegal(id, {
			cwd: normalized_cwd,
			node_modules: vite.normalizePath(path.resolve('node_modules')),
			server: vite.normalizePath(path.join(normalized_lib, 'server'))
		})
	) {
		const relative = normalize_id(id, normalized_lib, normalized_cwd);
		throw new Error(`Cannot import ${relative} into client-side code`);
	}
}
```

Running my script with the `TEST` env var set to `true` bypasses this piece of code: `TEST=true npx vite-node scripts/clear-db.ts`

🤷🏻‍♂️

Would love to know if there's a better way to go about it, but for now at least it works.

**Reply by david-plugge** [reply]
Yes there is, you need to tell vite that everything should be transformed as ssr code:
```bash
vite-node --options.transformMode.ssr='/.*/' src/lib/util/whatever.ts
```
